files/abuseipdb.local

```diff
@@ -0,0 +1,74 @@
+#   tracking, and isolation for IP abuse reporting.
+#
+# Key Features:
+#   - Isolated AbuseIPDB SQLite database to track reports independently.
+#   - Dual API calls: /v2/check and /v2/report for efficient reporting.
+#   - Optimizes AbuseIPDB daily API usage by reducing unnecessary report calls
+#   - Supports norestored=1 to avoid duplicate reports on restart.
+#   - Maintains a separate, long-term persistent database of banned IPs to ensure accurate tracking
+#     and avoid reliance on Fail2Ban’s bantimes or ban state, even if Fail2Ban’s
+#     internal ban management becomes inconsistent over time.
+#   - Customizable report comments to avoid leaking sensitive info.
+#
+# Example 'jail' configuration in 'jail.local' to prevent leaking sensitive information in AbuseIPDB reports:
+#   [nginx-botsearch]
+#   enabled    = true
+#   logpath    = /var/log/nginx/*.log
+#   port       = http,https
+#   backend    = polling
+#   tp_comment = Fail2Ban - NGINX bad requests 400-401-403-404-444, high level vulnerability scanning
+#   maxretry   = 3
+#   findtime   = 1d
+#   bantime    = 7200
+#   action     = %(action_mwl)s
+#                %(action_abuseipdb)s[matches="%(tp_comment)s", abuseipdb_apikey="YOUR_API_KEY", abuseipdb_category="21,15", bantime="%(bantime)s"]
+
+
+[Definition]
+# Option:  norestored
+######################
+# Notes.:  Ensure norestored is set to 0
+#          We control this at the script level to provide users with more control over how restored tickets are handled.
+#          We will also be able to log all triggered events by norestored
+#          Do not modify this value directly. Instead, adjust 'BYPASS_FAIL2BAN' below as needed.
+# norestored = 0
+
+
+# Option:  User defined settings
+######################
+# Notes.:  * Path to AbuseIPDB SQLite database used by the action script.
+#          * Path to the log file where actions and events are recorded by the action script.
+#          * Rely on Fail2Ban for restarts (0) or completely isolate it by bypassing Fail2Ban (1)
+#          ! Bypassing Fail2Ban on restarts (BYPASS_FAIL2BAN = 1) can overhelm your server and AbuseIPDB API on restarts.
+#          ! SET 1 if you want to completely isolate from Fail2Ban and rely solely on the AbuseIPDB SQLite database for reporting on restart.
+# SQLITE_DB = "/var/lib/fail2ban/abuseipdb/fail2ban_abuseipdb"
+# LOG_FILE = "/var/log/abuseipdb/abuseipdb.log"
+# BYPASS_FAIL2BAN = 0
+
+
+# Option:  actionstart
+######################
+# Notes.:  Uncomment and leave as-is
+# actionstart = nohup /etc/fail2ban/action.d/fail2ban-abuseipdb.sh \
+#     "--actionstart" "<SQLITE_DB>" "<LOG_FILE>" &
+
+
+# Option:  actionban
+######################
+# Notes.:  Uncomment and leave as-is
+# actionban = /etc/fail2ban/action.d/fail2ban-abuseipdb.sh \
+#     "<abuseipdb_apikey>" "<matches>" "<ip>" "<abuseipdb_category>" "<bantime>" "<restored>" "<BYPASS_FAIL2BAN>" "<SQLITE_DB>" "<LOG_FILE>"
```

Comment:
This will leak the API key in the `ps` output which could be an issue on a web server with SSH users. The usual workaround is to pass the secret in the environment but I don't know off the top of my head how to do that here.

Reply:
Well, it is questionable (at least if used in fail2ban) - it is running as root, so basically possible with ps & co for root or sudoers only.
But they are anyway able to find the key in fail2ban configuration files, so it doesn't matter.
Regarding the env-vars, fail2ban can supply arguments as variables to shell commands of action, but at the moment only if the argument contains some special character. At the moment, there is no possibility to force it for plain arguments (what, I think, the api-key is).
Maybe we shall indeed implement some option, to enforce then vars usage.
Alternatively, one could set the var in env of the systemd-unit and use that var here, but... it is like using a sledge-hammer to crack a nut, in my opinion for abovementioned reason.

Reply:
> Well, it is questionable (at least if used in fail2ban) - it is running as root, so basically possible with ps & co for root or sudoers only.

The full process list is "usually" still available to all users. (I use quotes because I don't know what every distro is doing.) On linux, you can mount `/proc` with `hidepid=1` to prevent other users from seeing root processes, but it's not the kernel default. My regular user can see the argument to all root processes, e.g.

```
$ ps aux | grep root
...
root     43667  0.2  0.0 106168 55616 pts/4    S+   08:22   0:00 emacs --no-window-system hello-world.txt
```

Ref: https://www.redhat.com/en/blog/hidepid-linux-hide-pid